File: src/lib/commands/activity.ts
import { tryInvoke, type CommandPromise } from ".";

export type Activity = {
	id: number;
	name: string;
	points_per_second: number;
};

export async function getAllActivities(): CommandPromise<Activity[]> {
	return tryInvoke("get_all_activities");
}

export async function createActivity(
	name: string,
	pointsPerSecond: number
): CommandPromise<Activity> {
	return tryInvoke("create_activity", { name, pointsPerSecond });
}

export async function deleteActivity(id: number): CommandPromise<void> {
	return tryInvoke("delete_activity", { id });
}

export async function editActivity(
	id: number,
	name: string,
	pointsPerSecond: number
): CommandPromise<Activity> {
	return tryInvoke("edit_activity", { id, name, pointsPerSecond });
}

File: src-tauri/src/commands/activity.rs
use crate::{
    database::{models::Activity, Db},
    types::{ActivityError, CommandError},
};
use tauri::State;

#[tauri::command]
pub async fn get_all_activities(db: State<'_, Db>) -> Result<Vec<Activity>, CommandError> {
    log::debug!("Getting all activities");
    let activities = Activity::get_all(&db).await?;
    log::debug!("Got {} activities successfully", activities.len());

    Ok(activities)
}

#[tauri::command]
pub async fn create_activity(
    db: State<'_, Db>,
    name: String,
    points_per_second: i8,
) -> Result<Activity, CommandError> {
    let name = name.trim().chars().take(30).collect::<String>();
    if name.is_empty() {
        return Err(ActivityError::EmptyName.into());
    }

    let points_per_second = points_per_second.clamp(-10, 10);

    log::debug!("Creating activity: {}", name);
    let activity = Activity::create(&db, name, points_per_second).await?;
    log::debug!("Created activity '{}' successfully", activity.name);

    Ok(activity)
}

#[tauri::command]
pub async fn delete_activity(db: State<'_, Db>, id: i64) -> Result<(), CommandError> {
    log::debug!("Deleting activity: {}", id);
    Activity::delete(&db, id).await?;
    log::debug!("Deleted activity '{}' successfully", id);

    Ok(())
}

#[tauri::command]
pub async fn edit_activity(
    db: State<'_, Db>,
    id: i64,
    name: String,
    points_per_second: i8,
) -> Result<Activity, CommandError> {
    let name = name.trim().chars().take(30).collect::<String>();
    if name.is_empty() {
        return Err(ActivityError::EmptyName.into());
    }

    let points_per_second = points_per_second.clamp(-10, 10);

    log::debug!("Editing activity: {}", id);
    let activity = Activity::edit(&db, id, name, points_per_second).await?;
    log::debug!("Edited activity '{}' successfully", id);

    Ok(activity)
}
